In [1]:
#importing neccessary libraries

import pandas as pd
import numpy as np

from sklearn.cluster import KMeans

import json 
from geopy.geocoders import Nominatim 

!conda install -c conda-forge folium=0.5.0 --yes
import folium

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    certifi-2020.4.5.2         |   py36h9f0ad1d_0         152 KB  conda-forge
    ca-certificates-2020.4.5.2 |       hecda079_0         147 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    ------------------------------------------------------------
                       

# all necessary libraries have been downloaded


# colllating data from Wikipedia

In [2]:
#scrapping data from url

url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

dfs = pd.read_html(url, header=0) # set header with row #0


# creating the dataframe from the requisite table 

In [3]:
#creating the required dataframe

df = dfs[0]
indexNames = df[ df['Borough'] == 'Not assigned' ].index
# Delete these row indexes from dataFrame
df.drop(indexNames , inplace=True)

# cleaning data with required index

index = range(1,df.shape[0]+1,1)
df["index"] = index
df.set_index('index', inplace = True)

df

Postal Code           Borough  \
index                                 
1             M3A        North York   
2             M4A        North York   
3             M5A  Downtown Toronto   
4             M6A        North York   
5             M7A  Downtown Toronto   
6             M9A         Etobicoke   
7             M1B       Scarborough   
8             M3B        North York   
9             M4B         East York   
10            M5B  Downtown Toronto   
11            M6B        North York   
12            M9B         Etobicoke   
13            M1C       Scarborough   
14            M3C        North York   
15            M4C         East York   
16            M5C  Downtown Toronto   
17            M6C              York   
18            M9C         Etobicoke   
19            M1E       Scarborough   
20            M4E      East Toronto   
21            M5E  Downtown Toronto   
22            M6E              York   
23            M1G       Scarborough   
24            M4G         East York   
25            M5G  Downtown Toronto   
26            M6G  Downtown Toronto   
27            M1H       Scarborough   
28            M2H        North York   
29            M3H        North York   
30            M4H         East York   
...           ...               ...   
74            M4R   Central Toronto   
75            M5R   Central Toronto   
76            M6R      West Toronto   
77            M7R       Mississauga   
78            M9R         Etobicoke   
79            M1S       Scarborough   
80            M4S   Central Toronto   
81            M5S  Downtown Toronto   
82            M6S      West Toronto   
83            M1T       Scarborough   
84            M4T   Central Toronto   
85            M5T  Downtown Toronto   
86            M1V       Scarborough   
87            M4V   Central Toronto   
88            M5V  Downtown Toronto   
89            M8V         Etobicoke   
90            M9V         Etobicoke   
91            M1W       Scarborough   
92            M4W  Downtown Toronto   
93            M5W  Downtown Toronto   
94            M8W         Etobicoke   
95            M9W         Etobicoke   
96            M1X       Scarborough   
97            M4X  Downtown Toronto   
98            M5X  Downtown Toronto   
99            M8X         Etobicoke   
100           M4Y  Downtown Toronto   
101           M7Y      East Toronto   
102           M8Y         Etobicoke   
103           M8Z         Etobicoke   

                                            Neighborhood  
index                                                     
1                                              Parkwoods  
2                                       Victoria Village  
3                              Regent Park, Harbourfront  
4                       Lawrence Manor, Lawrence Heights  
5            Queen's Park, Ontario Provincial Government  
6                Islington Avenue, Humber Valley Village  
7                                         Malvern, Rouge  
8                                              Don Mills  
9                        Parkview Hill, Woodbine Gardens  
10                              Garden District, Ryerson  
11                                             Glencairn  
12     West Deane Park, Princess Gardens, Martin Grov...  
13                Rouge Hill, Port Union, Highland Creek  
14                                             Don Mills  
15                                      Woodbine Heights  
16                                        St. James Town  
17                                    Humewood-Cedarvale  
18     Eringate, Bloordale Gardens, Old Burnhamthorpe...  
19                     Guildwood, Morningside, West Hill  
20                                           The Beaches  
21                                           Berczy Park  
22                                   Caledonia-Fairbanks  
23                                                Woburn  
24                                               Leaside  
25            

In [4]:
##number of rows in the dataframe

print("The number of rows in the dataframe are ",df.shape[0])

The number of rows in the dataframe are  103


# getting location data

In [241]:
# The code was removed by Watson Studio for sharing.

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [6]:
merged_data = df.join(df_loc.set_index('Postal Code'), on='Postal Code')


# merging location data and wikipedia table

In [7]:
merged_data.head()

Postal Code           Borough  \
index                                 
1             M3A        North York   
2             M4A        North York   
3             M5A  Downtown Toronto   
4             M6A        North York   
5             M7A  Downtown Toronto   

                                      Neighborhood   Latitude  Longitude  
index                                                                     
1                                        Parkwoods  43.753259 -79.329656  
2                                 Victoria Village  43.725882 -79.315572  
3                        Regent Park, Harbourfront  43.654260 -79.360636  
4                 Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
5      Queen's Park, Ontario Provincial Government  43.662301 -79.389494

In [8]:
address = 'Toronto'

geolocator = Nominatim(user_agent="my-application")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


# plotting map details for toronto city

In [10]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhoods in zip(merged_data['Latitude'], merged_data['Longitude'], merged_data['Borough'], merged_data['Neighborhood']):
    label = '{}, {}'.format(neighborhoods, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_toronto)  
    
map_toronto

# Focussing on those neighbourhood which have the string toronto

In [13]:
neighbourhood_list = merged_data['Borough'].unique()
    


array(['North York', 'Downtown Toronto', 'Etobicoke', 'Scarborough',
       'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

In [18]:
toronto_list = []

for i in neighbourhood_list :
    input_string = i.lower()
    x = input_string.find('toronto')
    if x != -1:
        toronto_list.append(i)

toronto_list

['Downtown Toronto', 'East Toronto', 'West Toronto', 'Central Toronto']

In [28]:
df_toronto = merged_data[merged_data['Borough'].isin(toronto_list)]




In [29]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighborhoods in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhoods, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_toronto)  
    
map_toronto

In [242]:
# The code was removed by Watson Studio for sharing.

# exploring venue details for the required neighbourhods

In [62]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhoods in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Postal Code'], df_toronto['Borough'], df_toronto['Neighborhood']):
    
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
  
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhoods,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

# Creating venue dataframe

In [63]:
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Postal_Code', 'Borough', 'Neighborhoods', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head()


(1613, 9)


Postal_Code           Borough              Neighborhoods  BoroughLatitude  \
0         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   
1         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   
2         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   
3         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   
4         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   

   BoroughLongitude               VenueName  VenueLatitude  VenueLongitude  \
0        -79.360636        Roselle Desserts      43.653447      -79.362017   
1        -79.360636           Tandem Coffee      43.653559      -79.361809   
2        -79.360636  Cooper Koo Family YMCA      43.653249      -79.358008   
3        -79.360636     Body Blitz Spa East      43.654735      -79.359874   
4        -79.360636         Corktown Common      43.655618      -79.356211   

         VenueCategory  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4                 Park

In [94]:
neighborhood_group = venues_df.groupby('Neighborhoods')
neighborhood_group

In [71]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 239 uniques categories.


# Creating one hot table for neighbourhoods and venue category

In [166]:
toronto_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

#add neighborhood column back to dataframe
toronto_onehot['Neighborhoods'] = venues_df['Neighborhoods'] 

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head(39)


Neighborhoods  Afghan Restaurant  Airport  Airport Food Court  \
0   Regent Park, Harbourfront                  0        0                   0   
1   Regent Park, Harbourfront                  0        0                   0   
2   Regent Park, Harbourfront                  0        0                   0   
3   Regent Park, Harbourfront                  0        0                   0   
4   Regent Park, Harbourfront                  0        0                   0   
5   Regent Park, Harbourfront                  0        0                   0   
6   Regent Park, Harbourfront                  0        0                   0   
7   Regent Park, Harbourfront                  0        0                   0   
8   Regent Park, Harbourfront                  0        0                   0   
9   Regent Park, Harbourfront                  0        0                   0   
10  Regent Park, Harbourfront                  0        0                   0   
11  Regent Park, Harbourfront                  0        0                   0   
12  Regent Park, Harbourfront                  0        0                   0   
13  Regent Park, Harbourfront                  0        0                   0   
14  Regent Park, Harbourfront                  0        0                   0   
15  Regent Park, Harbourfront                  0        0                   0   
16  Regent Park, Harbourfront                  0        0                   0   
17  Regent Park, Harbourfront                  0        0                   0   
18  Regent Park, Harbourfront                  0        0                   0   
19  Regent Park, Harbourfront                  0        0                   0   
20  Regent Park, Harbourfront                  0        0                   0   
21  Regent Park, Harbourfront                  0        0                   0   
22  Regent Park, Harbourfront                  0        0                   0   
23  Regent Park, Harbourfront                  0        0                   0   
24  Regent Park, Harbourfront                  0        0                   0   
25  Regent Park, Harbourfront                  0        0                   0   
26  Regent Park, Harbourfront                  0        0                   0   
27  Regent Park, Harbourfront                  0        0                   0   
28  Regent Park, Harbourfront                  0        0                   0   
29  Regent Park, Harbourfront                  0        0                   0   
30  Regent Park, Harbourfront                  0        0                   0   
31  Regent Park, Harbourfront                  0        0                   0   
32  Regent Park, Harbourfront                  0        0                   0   
33  Regent Park, Harbourfront                  0        0                   0   
34  Regent Park, Harbourfront                  0        0                   0   
35  Regent Park, Harbourfront                  0        0                   0   
36  Regent Park, Harbourfront                  0        0                   0   
37  Regent Park, Harbourfront                  0        0                   0   
38  Regent Park, Harbourfront                  0        0                   0   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0              0               0                0                 0   
1              0               0                0                 0   
2              0               0                0                 0   
3              0               0                0                 0   
4              0               0                0                 0   
5              0               0                0                 0   
6              0               0                0                 0   
7              0               0                0                 0   
8              0               0                0                 0   
9              0               0                0                

In [128]:
neighborhood_group = toronto_onehot.groupby('Neighborhoods').mean().reset_index()



In [130]:
neighborhood_group.head()
neighborhood_group.shape


(39, 240)

# Top venues for each neighbourhood

In [243]:
num_top_venues = 4

for hood in neighborhood_group['Neighborhoods']:
    print("----"+hood+"----")
    temp =neighborhood_group[neighborhood_group['Neighborhoods'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
         venue  freq
0  Coffee Shop  0.09
1  Cheese Shop  0.03
2       Bakery  0.03
3   Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1     Coffee Shop  0.09
2       Nightclub  0.09
3  Breakfast Spot  0.09


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
           venue  freq
0    Pizza Place  0.06
1  Auto Workshop  0.06
2     Comic Shop  0.06
3     Restaurant  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3          Boutique  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1      Sandwich Place  0.06
2  Italian Restaurant  0.06
3                Café  0.05


----Christie----
           venue  freq
0  Grocery Store  0.24
1        

# top 10 venues for each neighbourhood

In [221]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]




num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhoods']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhoods'] = neighborhood_group['Neighborhoods']


for ind in np.arange(neighborhood_group.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(neighborhood_group.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()
    

Neighborhoods 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...                  Park   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar            Restaurant           Cheese Shop   
1           Coffee Shop        Breakfast Spot             Nightclub   
2           Pizza Place    Light Rail Station            Smoke Shop   
3       Airport Service      Airport Terminal       Harbor / Marina   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Seafood Restaurant              Beer Bar                Bakery   
1             Pet Store               Stadium         Burrito Place   
2               Brewery         Burrito Place        Farmers Market   
3              Boutique   Rental Car Location                 Plane   
4       Thai Restaurant          Burger Joint   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café     French Restaurant     Basketball Stadium  
1            Restaurant          Climbing Gym  Performing Arts Venue  
2  Fast Food Restaurant               Butcher             Restaurant  
3         Boat or Ferry                   Bar       Sculpture Garden  
4      Department Store           Salad Place        Bubble Tea Shop

# clustering and finding hidden patterns (k-means, unsupervised) k = 3 

In [222]:
# set number of clusters
kclusters = 3

neighborhood_group_clustering = neighborhood_group.drop('Neighborhoods', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(neighborhood_group_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:39] 


array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1,
       1, 1, 1, 1, 0, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [251]:
#add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)


toronto_merged = merged_data
toronto_merged

#merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
final_df = neighborhoods_venues_sorted.join(toronto_merged.set_index('Neighborhood'), on='Neighborhoods')




In [252]:
final_df.head()

Cluster Labels                                      Neighborhoods  \
0               1                                        Berczy Park   
1               1       Brockton, Parkdale Village, Exhibition Place   
2               1  Business reply mail Processing Centre, South C...   
3               1  CN Tower, King and Spadina, Railway Lands, Har...   
4               1                                 Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop          Cocktail Bar            Restaurant   
1                  Café           Coffee Shop        Breakfast Spot   
2                  Park           Pizza Place    Light Rail Station   
3        Airport Lounge       Airport Service      Airport Terminal   
4           Coffee Shop    Italian Restaurant        Sandwich Place   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0           Cheese Shop    Seafood Restaurant              Beer Bar   
1             Nightclub             Pet Store               Stadium   
2            Smoke Shop               Brewery         Burrito Place   
3       Harbor / Marina              Boutique   Rental Car Location   
4                  Café       Thai Restaurant          Burger Joint   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                Bakery                  Café     French Restaurant   
1         Burrito Place            Restaurant          Climbing Gym   
2        Farmers Market  Fast Food Restaurant               Butcher   
3                 Plane         Boat or Ferry                   Bar   
4   Japanese Restaurant      Department Store           Salad Place   

  10th Most Common Venue Postal Code           Borough   Latitude  Longitude  
0     Basketball Stadium         M5E  Downtown Toronto  43.644771 -79.373306  
1  Performing Arts Venue         M6K      West Toronto  43.636847 -79.428191  
2             Restaurant         M7Y      East Toronto  43.662744 -79.321558  
3       Sculpture Garden         M5V  Downtown Toronto  43.628947 -79.394420  
4        Bubble Tea Shop         M5G  Downtown Toronto  43.657952 -79.387383

# plotting on map

In [225]:
import matplotlib.cm as cm
import matplotlib.colors as colors

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
colors_array

array([[5.00000000e-01, 0.00000000e+00, 1.00000000e+00, 1.00000000e+00],
       [5.03921569e-01, 9.99981027e-01, 7.04925547e-01, 1.00000000e+00],
       [1.00000000e+00, 1.22464680e-16, 6.12323400e-17, 1.00000000e+00]])

In [226]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

       
rainbow

['#8000ff', '#80ffb4', '#ff0000']

In [212]:
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(final_df['Latitude'], final_df['Longitude'], final_df['Neighborhoods'], final_df['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    print(cluster)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

1
1
1
1
1
1
1
1
1
1
1
1
0
1
1
1
1
1
0
1
0
1
1
1
1
1
0
2
1
1
1
1
1
1
1
1
1
1
1


# cluster wise tables

In [236]:
final_df.loc[final_df['Cluster Labels'] == 0, final_df.columns[[1] + list(range(2, final_df.shape[1]))]]

Neighborhoods 1st Most Common Venue  \
12  Forest Hill North & West, Forest Hill Road Park                  Park   
18                                    Lawrence Park                  Park   
20                      Moore Park, Summerhill East                Lawyer   
26                                         Rosedale                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12         Jewelry Store                 Trail      Sushi Restaurant   
18           Swim School              Bus Line         Deli / Bodega   
20                 Trail            Restaurant                  Park   
26            Playground                 Trail          Cupcake Shop   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12                 Dance Studio   Dumpling Restaurant            Donut Shop   
18  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
20        General Entertainment     German Restaurant            Donut Shop   
26          Dumpling Restaurant            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
12      Doner Restaurant               Dog Run    Distribution Center   
18      Doner Restaurant               Dog Run    Distribution Center   
20      Doner Restaurant               Dog Run    Distribution Center   
26               Dog Run   Distribution Center         Discount Store   

   Postal Code           Borough   Latitude  Longitude  
12         M5P   Central Toronto  43.696948 -79.411307  
18         M4N   Central Toronto  43.728020 -79.388790  
20         M4T   Central Toronto  43.689574 -79.383160  
26         M4W  Downtown Toronto  43.679563 -79.377529

# conclusion ; cluster 0 consistes of those neighbourhoods where there are more outdoor activity spots such as parks/playgrounds/dog run/trail and few restaurants

In [237]:
final_df.loc[final_df['Cluster Labels'] == 1, final_df.columns[[1] + list(range(2, final_df.shape[1]))]]

Neighborhoods 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place                  Café   
2   Business reply mail Processing Centre, South C...                  Park   
3   CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville        Sandwich Place   
9                                    Davisville North                  Park   
10                       Dufferin, Dovercourt Village                Bakery   
11             First Canadian Place, Underground city           Coffee Shop   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
15                      High Park, The Junction South    Mexican Restaurant   
16                     India Bazaar, The Beaches West  Fast Food Restaurant   
17          Kensington Market, Chinatown, Grange Park                  Café   
19                           Little Portugal, Trinity                   Bar   
21                  North Toronto West, Lawrence Park           Coffee Shop   
22                             Parkdale, Roncesvalles             Gift Shop   
23        Queen's Park, Ontario Provincial Government           Coffee Shop   
24                          Regent Park, Harbourfront           Coffee Shop   
25                           Richmond, Adelaide, King           Coffee Shop   
28                                 Runnymede, Swansea           Pizza Place   
29                                     St. James Town           Coffee Shop   
30                        St. James Town, Cabbagetown           Coffee Shop   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District                  Café   
33  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
34                The Annex, North Midtown, Yorkville                  Café   
35                                        The Beaches                 Trail   
36                       The Danforth West, Riverdale      Greek Restaurant   
37           Toronto Dominion Centre, Design Exchange           Coffee Shop   
38                     University of Toronto, Harbord                  Café   

   2nd Most Common Venue      3rd Most Common Venue  \
0           Cocktail Bar                 Restaurant   
1            Coffee Shop             Breakfast Spot   
2            Pizza Place         Light Rail Station   
3        Airport Service           Airport Terminal   
4     Italian Restaurant             Sandwich Place   
5                   Café                       Park   
6       Sushi Restaurant        Japanese Restaurant   
7             Restaurant                       Café   
8           Dessert Shop                Coffee Shop   
9         Sandwich Place           Department Store   
10              Pharmacy              Grocery Store   
11                  Café                      Hotel   
13           Coffee Shop  Middle Eastern Restaurant   
14              Aquarium                       Café   
15       Thai Restaurant                       Café   
16        Sandwich Place                       Park   
17           Coffee Shop         Mexican Restaurant   
19            Restaurant                       Café   
21        Clothing Store                 Bagel Shop   
22        Breakfast Spot                 Restaurant   
23      Sushi Restaurant                Yoga Studio   
24                Bakery                        Pub   
25       

# conclusion ; cluster 1 consistes of those neighbourhoods where there are numerous restauraunts/bars/cafe/eateries

In [238]:
final_df.loc[final_df['Cluster Labels'] == 2, final_df.columns[[1] + list(range(2, final_df.shape[1]))]]

Neighborhoods    1st Most Common Venue 2nd Most Common Venue  \
27      Roselawn  Health & Beauty Service                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
27           Yoga Studio         Deli / Bodega     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
27  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue Postal Code          Borough  \
27      Doner Restaurant                Dog Run         M5N  Central Toronto   

     Latitude  Longitude  
27  43.711695 -79.416936

# conclusion ; cluster 2 consistes of one neighbourhood and seems to be an outlier.

# Final Conclustion ; There were only 39 neighbourhoods to cluster across 240 categories. Hence the results were not optimun in nature. Probably the number of categories could have been reduced to improve the results or the number of neighbourhoods should increase.